File: package-python/python/perpetual/policy.py
"""Policy learning via Inverse Propensity Weighting (IPW / AIPW).

Implements the Athey & Wager (2021) policy-learning framework using
gradient boosting.  The learned policy assigns treatment when the
boosted score is positive.

Two modes are supported:

* **IPW** — Standard Inverse Propensity Weighting.
* **AIPW** (Augmented / Doubly Robust) — Reduces variance by
  incorporating a baseline outcome model.
"""

from typing import Optional

import numpy as np
from typing_extensions import Self

from perpetual.booster import PerpetualBooster
from perpetual.perpetual import PolicyObjective

# ---------------------------------------------------------------------------
# Custom objective helpers
# ---------------------------------------------------------------------------


# ---------------------------------------------------------------------------
# PolicyLearner
# ---------------------------------------------------------------------------


class PolicyLearner:
    r"""Policy learner via Inverse Propensity Weighting.

    Learns a treatment-assignment policy :math:`\\pi(X)` that maximizes
    expected reward using the Athey & Wager (2021) policy-learning
    framework.

    The learned policy assigns :math:`W = 1` when the boosted
    score :math:`F(X) > 0`.

    Parameters
    ----------
    budget : float, default=0.5
        Fitting budget forwarded to ``PerpetualBooster``.
    mode : str, default="ipw"
        ``"ipw"`` for standard Inverse Propensity Weighting or
        ``"aipw"`` for Augmented (Doubly Robust) IPW.
    propensity_budget : float, optional
        Separate budget for the propensity model.  If ``None``, defaults to
        ``budget``.  Only used when ``propensity`` is not supplied to
        :meth:`fit`.
    **kwargs
        Additional keyword arguments forwarded to ``PerpetualBooster``.

    Attributes
    ----------
    feature_importances_ : ndarray of shape (n_features,)
        Feature importances from the policy model.

    Examples
    --------
    >>> from perpetual.policy import PolicyLearner
    >>> import numpy as np
    >>> n = 500
    >>> X = np.random.randn(n, 5)
    >>> w = np.random.binomial(1, 0.5, n)
    >>> y = X[:, 0] * w + np.random.randn(n) * 0.5
    >>> pl = PolicyLearner(budget=0.3)
    >>> pl.fit(X, w, y)  # doctest: +SKIP
    >>> policy = pl.predict(X)  # doctest: +SKIP

    References
    ----------
    Athey, S., & Wager, S. (2021). *Policy learning with observational data*.
    Econometrica, 89(1), 133-161.
    """

    def __init__(
        self,
        budget: float = 0.5,
        mode: str = "ipw",
        propensity_budget: Optional[float] = None,
        **kwargs,
    ):
        if mode not in ("ipw", "aipw"):
            raise ValueError(f"mode must be 'ipw' or 'aipw', got {mode!r}.")
        self.budget = budget
        self.mode = mode
        self.propensity_budget = propensity_budget or budget
        self._kwargs = kwargs
        self.feature_importances_: Optional[np.ndarray] = None

    def fit(
        self,
        X,
        w,
        y,
        propensity: Optional[np.ndarray] = None,
        mu_hat_1: Optional[np.ndarray] = None,
        mu_hat_0: Optional[np.ndarray] = None,
    ) -> Self:
        r"""Fit the policy learner.

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Covariate matrix.
        w : array-like of shape (n_samples,)
            Observed binary treatment assignment (0 or 1).
        y : array-like of shape (n_samples,)
            Observed outcome.
        propensity : array-like of shape (n_samples,), optional
            Estimated :math:`P(W=1|X)`.  If ``None``, a propensity model is
            fitted internally.
        mu_hat_1 : array-like of shape (n_samples,), optional
            Predicted outcome under treatment :math:`\hat{\mu}_1(X)`.
            Required when ``mode="aipw"``.
        mu_hat_0 : array-like of shape (n_samples,), optional
            Predicted outcome under control :math:`\hat{\mu}_0(X)`.
            Required when ``mode="aipw"``.

        Returns
        -------
        self
            Fitted estimator.
        """
        X_arr = np.asarray(X, dtype=float)
        w_arr = np.asarray(w, dtype=float)
        y_arr = np.asarray(y, dtype=float)

        # Propensity scores
        if propensity is not None:
            p_hat = np.asarray(propensity, dtype=float)
        else:
            prop_model = PerpetualBooster(
                budget=self.propensity_budget, objective="LogLoss", **self._kwargs
            )
            prop_model.fit(X_arr, w_arr)
            log_odds = prop_model.predict(X_arr)
            p_hat = 1.0 / (1.0 + np.exp(-log_odds))

        # Baseline outcome models for AIPW
        m1 = None
        m0 = None
        if self.mode == "aipw":
            if mu_hat_1 is not None and mu_hat_0 is not None:
                m1 = np.asarray(mu_hat_1, dtype=float)
                m0 = np.asarray(mu_hat_0, dtype=float)
            else:
                # Fit separate outcome models for treatment and control
                m1_model = PerpetualBooster(
                    budget=self.budget, objective="SquaredLoss", **self._kwargs
                )
                m0_model = PerpetualBooster(
                    budget=self.budget, objective="SquaredLoss", **self._kwargs
                )
                mask1 = w_arr == 1
                mask0 = w_arr == 0
                if mask1.any():
                    m1_model.fit(X_arr[mask1], y_arr[mask1])
                if mask0.any():
                    m0_model.fit(X_arr[mask0], y_arr[mask0])

                m1 = m1_model.predict(X_arr)
                m0 = m0_model.predict(X_arr)

        # Policy model via Rust PolicyObjective
        objective = PolicyObjective(
            treatment=w_arr.astype("uint8"),
            propensity=p_hat,
            mode=self.mode,
            mu_hat_1=m1,
            mu_hat_0=m0,
        )

        self._policy_model = PerpetualBooster(
            budget=self.budget, objective=objective, **self._kwargs
        )
        # Pass observed outcome y_arr. The definition of Gamma is handled internally
        # by the RustPolicyObjective, which implements pseudo_outcome using y.
        self._policy_model.fit(X_arr, y_arr)
        self.feature_importances_ = self._policy_model.feature_importances_

        return self

    def predict(self, X) -> np.ndarray:
        """Predict the optimal treatment assignment.

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Covariate matrix.

        Returns
        -------
        ndarray of shape (n_samples,)
            Binary treatment policy (1 = treat, 0 = do not treat).
        """
        scores = self.decision_function(X)
        return (scores > 0).astype(int)

    def decision_function(self, X) -> np.ndarray:
        """Return raw policy scores.

        Positive values indicate treatment is beneficial.

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Covariate matrix.

        Returns
        -------
        ndarray of shape (n_samples,)
            Raw boosted policy scores.
        """
        return self._policy_model.predict(X)

    def predict_proba(self, X) -> np.ndarray:
        """Predict probability that treatment is beneficial.

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Covariate matrix.

        Returns
        -------
        ndarray of shape (n_samples,)
            Probability of treatment being beneficial (sigmoid of score).
        """
        scores = self.decision_function(X)
        return 1.0 / (1.0 + np.exp(-scores))

File: src/rust/core/src/causal/policy.rs
//! Policy Gradient Objective
//!
//! An inverse-propensity-weighted (IPW) objective that learns a policy
//! maximizing expected reward via gradient boosting.
//!
//! Two modes are supported:
//!
//! * **IPW** – Standard Inverse Propensity Weighting.
//! * **AIPW** (Augmented IPW / Doubly Robust) – Reduces variance by
//!   incorporating a baseline outcome model $\hat{\mu}(X)$.
use crate::metrics::evaluation::Metric;
use crate::objective::ObjectiveFunction;
use serde::{Deserialize, Serialize};

/// Minimum propensity clip for numerical safety.
const PROPENSITY_CLIP: f64 = 1e-3;

/// Policy learning mode.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum PolicyMode {
    /// Standard Inverse Propensity Weighting.
    IPW,
    /// Augmented / Doubly Robust IPW with baseline outcome predictions.
    AIPW {
        /// Predicted baseline outcome $\hat{\mu}_1(X)$ for treated.
        mu_hat_1: Vec<f64>,
        /// Predicted baseline outcome $\hat{\mu}_0(X)$ for control.
        mu_hat_0: Vec<f64>,
    },
}

/// Policy Gradient Objective (Inverse Propensity Weighting).
///
/// Optimizes a policy score $F(x)$ to maximize the expected reward using
/// the Athey & Wager (2021) Policy Learning approach.
///
/// The learned policy assigns treatment $W = 1$ when $\sigma(F(x)) > 0.5$
/// (equivalently $F(x) > 0$).
///
/// For binary treatment $W \in \{0, 1\}$:
/// - The IPW pseudo-outcome for each sample is:
///   $$\Gamma_i = \frac{Y_i \cdot W_i}{p_i} - \frac{Y_i \cdot (1 - W_i)}{1 - p_i}$$
/// - The objective reduces to logistic regression weighted by $|\Gamma_i|$.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PolicyObjective {
    /// Observed binary treatment assignment (0 or 1).
    pub treatment: Vec<u8>,
    /// Estimated propensity score $P(W=1|X)$ for each sample.
    pub propensity: Vec<f64>,
    /// Policy mode: IPW or AIPW.
    pub mode: PolicyMode,
}

impl PolicyObjective {
    /// Create a new IPW `PolicyObjective`.
    ///
    /// * `treatment` - Binary treatment assignments.
    /// * `propensity` - Estimated $P(W=1|X)$ for each sample.
    pub fn new(treatment: Vec<u8>, propensity: Vec<f64>) -> Self {
        Self {
            treatment,
            propensity,
            mode: PolicyMode::IPW,
        }
    }

    /// Create a new AIPW (Doubly Robust) `PolicyObjective`.
    ///
    /// * `treatment` - Binary treatment assignments.
    /// * `propensity` - Estimated $P(W=1|X)$ for each sample.
    /// * `mu_hat_1` - Predicted baseline outcome $\hat{\mu}_1(X)$ for treated.
    /// * `mu_hat_0` - Predicted baseline outcome $\hat{\mu}_0(X)$ for control.
    pub fn new_aipw(treatment: Vec<u8>, propensity: Vec<f64>, mu_hat_1: Vec<f64>, mu_hat_0: Vec<f64>) -> Self {
        Self {
            treatment,
            propensity,
            mode: PolicyMode::AIPW { mu_hat_1, mu_hat_0 },
        }
    }

    /// Compute the IPW or AIPW pseudo-outcome $\Gamma_i$ for each sample.
    ///
    /// $$\Gamma_i^{IPW} = \frac{Y_i W_i}{p_i} - \frac{Y_i (1-W_i)}{1-p_i}$$
    ///
    /// $$\Gamma_i^{AIPW} = (\hat{\mu}_1 - \hat{\mu}_0)
    ///     + \frac{W_i (Y_i - \hat{\mu}_1)}{p_i}
    ///     - \frac{(1-W_i)(Y_i - \hat{\mu}_0)}{1-p_i}$$
    fn pseudo_outcome(&self, y: &[f64]) -> Vec<f64> {
        let n = y.len();
        let mut gamma = Vec::with_capacity(n);

        for i in 0..n {
            let w = self.treatment[i] as f64;
            let p = self.propensity[i].clamp(PROPENSITY_CLIP, 1.0 - PROPENSITY_CLIP);
            let yi = y[i];

            let g = match &self.mode {
                PolicyMode::IPW => {
                    // Gamma = Y*W/p - Y*(1-W)/(1-p)
                    yi * w / p - yi * (1.0 - w) / (1.0 - p)
                }
                PolicyMode::AIPW { mu_hat_1, mu_hat_0 } => {
                    let m1 = mu_hat_1[i];
                    let m0 = mu_hat_0[i];

                    // AIPW score = (m1 - m0) + W(Y-m1)/p - (1-W)(Y-m0)/(1-p)
                    let term1 = m1 - m0;
                    let term2 = w * (yi - m1) / p;
                    let term3 = (1.0 - w) * (yi - m0) / (1.0 - p);

                    term1 + term2 - term3
                }
            };
            gamma.push(g);
        }
        gamma
    }
}

impl ObjectiveFunction for PolicyObjective {
    fn loss(&self, y: &[f64], yhat: &[f64], _sample_weight: Option<&[f64]>, _group: Option<&[u64]>) -> Vec<f32> {
        // Weighted logistic loss: -|Gamma| * [ target * log(sigma) + (1-target) * log(1-sigma) ]
        let gamma = self.pseudo_outcome(y);
        gamma
            .iter()
            .zip(yhat.iter())
            .map(|(g, score)| {
                let sigma = 1.0 / (1.0 + (-score).exp());
                let target = if *g >= 0.0 { 1.0 } else { 0.0 };
                let w = g.abs();
                let loss = -(target * sigma.max(1e-15).ln() + (1.0 - target) * (1.0 - sigma).max(1e-15).ln());
                (w * loss) as f32
            })
            .collect()
    }

    fn gradient(
        &self,
        y: &[f64],
        yhat: &[f64],
        _sample_weight: Option<&[f64]>,
        _group: Option<&[u64]>,
    ) -> (Vec<f32>, Option<Vec<f32>>) {
        // Transform to weighted binary classification via pseudo-outcomes.
        //
        // Gamma_i > 0  →  target = 1, weight = |Gamma_i|  (treat is beneficial)
        // Gamma_i < 0  →  target = 0, weight = |Gamma_i|  (treat is harmful)
        //
        // Gradient:  w_i * (sigma(F) - target_i)
        // Hessian:   w_i * sigma(F) * (1 - sigma(F))
        let n = y.len();
        let gamma = self.pseudo_outcome(y);
        let mut grad = Vec::with_capacity(n);
        let mut hess = Vec::with_capacity(n);

        for i in 0..n {
            let score = yhat[i];
            let sigma = 1.0 / (1.0 + (-score).exp());

            let target = if gamma[i] >= 0.0 { 1.0 } else { 0.0 };
            let weight = gamma[i].abs();

            let g = weight * (sigma - target);
            let h = weight * sigma * (1.0 - sigma);

            grad.push(g as f32);
            hess.push(h as f32);
        }

        (grad, Some(hess))
    }

    fn initial_value(&self, _y: &[f64], _sample_weight: Option<&[f64]>, _group: Option<&[u64]>) -> f64 {
        0.0
    }

    fn default_metric(&self) -> Metric {
        Metric::LogLoss
    }
}
